laurent22/joplin
# Clipper, Desktop: Resolves #6247: Clipper unable to pull and store PDFs

As discussed in #6247 this PR has 2 major objective:

- Clipper identifies embedded pdfs in the page and saves them as resources.
- When we open a pdf file in browser and run clipper, it correctly saves the pdf file as a note.

## How to check it out

- Open the links [embedded pdf](https://pdfobject.com/) and [pdf file](http://www.doiserbia.nb.rs/img/doi/1450-5339/2008/1450-53390801039K.pdf) in the browser.
- Open the clipper plugin on each page and select any `clip page` option.
- Check the newly created note which must contain the pdfs. 
- If you chose a HTML option you will see a `<a>` link. You wont be able to preview it from the note itself but your pdf has been saved and clicking on the link will open the pdf. If you chose a Markdown option you should be able to see the preview of the pdf as well.

## Screenshots 

### Original page with embedded PDF
![Screenshot 2022-04-09 at 6 55 07 PM](https://user-images.githubusercontent.com/44570278/162578425-4d7c3a8a-baeb-480d-adc4-48c97d5bf12b.png)

### Clipped as Markdown
![Screenshot 2022-04-09 at 6 54 52 PM](https://user-images.githubusercontent.com/44570278/162578437-6b1fda91-4a30-444f-8840-9b52325d3e49.png)

### Clipped as HTML
![Screenshot 2022-04-09 at 6 56 14 PM](https://user-images.githubusercontent.com/44570278/162578431-1325a388-6015-4ee8-9c19-0a69d4df84ad.png)

### A raw .pdf page clipped into a note (markdown clipped)
![Screenshot 2022-04-09 at 6 57 58 PM](https://user-images.githubusercontent.com/44570278/162578439-dcd3f263-7afd-434a-84de-d8c12f181e82.png)


## Implementation notes

### For saving as HTML

We check for `<embed>` and `<object>` tags for pdf embeds since this is the standard way of web to embed them, and after converting them to resources we replace those tags to `<a>` links since `<embed>` and `<object>` tags are disabled in the renderer due to security reasons (CVE-2020-15930). 

### For saving as Markdown

We covert `<embed>`, `<object>` tags to its appropriate alternative in md i.e  into`[name](url)`. But this causes an issue since this conversion happens before their resources are created, and since in md there's no way to distinguish a normal link and a pdf embed this raises the question of which link tags should be converted to resources and which ones are to be left as it is. We cannot just convert all pdf links (i.e ending with `.pdf` ext) into a resource since it might just happen to be a normal hyperlink in the original page. As a solution I decided to append a `_pdf:` marker in front of the pdf links temporarily that needs to converted to resources, later once all the resources are created we remove the marker and url and it gets replaced with the new internal links to the resources thus the marker is invisible to the end user or any other parts of the system.

### Clipping a raw `.pdf` file

To make this work we are checking the `content-type` of the page on clipper and if it happens to be of type `application/pdf` we are sending the body of the note as `<embed src="[page url]" type="application/pdf" />` instead of trying to parse the document any further.

## Unit tests

2 new unit tests are added to `Api.text.ts` of `lib` package.
- should extract media urls from body
- should create notes with pdf embeds

The last one almost covers the whole PR.
